snowflake/repository.go

```diff
@@ -131,23 +131,38 @@ func (repo *SnowflakeRepository) GetDataUsage(ctx context.Context, minTime time.
 			}
 		}()
 
-		arguments := make([]any, 0, 2)
-		arguments = append(arguments, minTime)
+		queryGen := func(startTime time.Time) (string, []any) {
+			strBuilder := strings.Builder{}
+			args := make([]any, 0, 3)
 
-		endTime := ""
+			// First query only the QUERY_HISTORY (to avoid a join without LIMIT)
+			strBuilder.WriteString("WITH history as (\n")
+			strBuilder.WriteString(`SELECT QUERY_HISTORY.QUERY_ID as QUERY_ID, QUERY_HISTORY.QUERY_TEXT as QUERY_TEXT, DATABASE_NAME, SCHEMA_NAME, QUERY_TYPE, SESSION_ID, QUERY_HISTORY.USER_NAME as USER_NAME, ROLE_NAME, EXECUTION_STATUS, START_TIME, END_TIME, TOTAL_ELAPSED_TIME, BYTES_SCANNED, BYTES_WRITTEN, BYTES_WRITTEN_TO_RESULT, ROWS_PRODUCED, ROWS_INSERTED, ROWS_UPDATED, ROWS_DELETED, ROWS_UNLOADED, CREDITS_USED_CLOUD_SERVICES FROM "SNOWFLAKE"."ACCOUNT_USAGE"."QUERY_HISTORY" WHERE START_TIME > ? `)
 
-		if maxTime != nil {
-			endTime = "AND START_TIME <= ?"
+			args = append(args, startTime)
 
-			arguments = append(arguments, *maxTime)
-		}
+			if maxTime != nil {
+				strBuilder.WriteString("AND START_TIME <= ? ")
+
+				args = append(args, *maxTime)
+			}
+
+			if repo.usageBatchSize > 0 {
+				strBuilder.WriteString(" ORDER BY START_TIME asc LIMIT ?")
+
+				args = append(args, repo.usageBatchSize)
+			}
+
+			strBuilder.WriteString(")")
 
-		query := fmt.Sprintf(`SELECT QUERY_HISTORY.QUERY_ID as QUERY_ID, QUERY_HISTORY.QUERY_TEXT as QUERY_TEXT, DATABASE_NAME, SCHEMA_NAME, QUERY_TYPE, SESSION_ID, QUERY_HISTORY.USER_NAME as USER_NAME, ROLE_NAME, EXECUTION_STATUS, START_TIME, END_TIME, TOTAL_ELAPSED_TIME, BYTES_SCANNED, BYTES_WRITTEN, BYTES_WRITTEN_TO_RESULT, ROWS_PRODUCED, ROWS_INSERTED, ROWS_UPDATED, ROWS_DELETED, ROWS_UNLOADED, CREDITS_USED_CLOUD_SERVICES, DIRECT_OBJECTS_ACCESSED, BASE_OBJECTS_ACCESSED, OBJECTS_MODIFIED, OBJECT_MODIFIED_BY_DDL, PARENT_QUERY_ID, ROOT_QUERY_ID
-FROM "SNOWFLAKE"."ACCOUNT_USAGE"."QUERY_HISTORY" INNER JOIN "SNOWFLAKE"."ACCOUNT_USAGE"."ACCESS_HISTORY" ON QUERY_HISTORY.QUERY_ID = ACCESS_HISTORY.QUERY_ID
-WHERE START_TIME >= ? %s`, endTime)
+			// THEN join with ACCESS_HISTORY
+			strBuilder.WriteString(`SELECT QUERY_HISTORY.QUERY_ID as QUERY_ID, QUERY_HISTORY.QUERY_TEXT as QUERY_TEXT, DATABASE_NAME, SCHEMA_NAME, QUERY_TYPE, SESSION_ID, QUERY_HISTORY.USER_NAME as USER_NAME, ROLE_NAME, EXECUTION_STATUS, START_TIME, END_TIME, TOTAL_ELAPSED_TIME, BYTES_SCANNED, BYTES_WRITTEN, BYTES_WRITTEN_TO_RESULT, ROWS_PRODUCED, ROWS_INSERTED, ROWS_UPDATED, ROWS_DELETED, ROWS_UNLOADED, CREDITS_USED_CLOUD_SERVICES, DIRECT_OBJECTS_ACCESSED, BASE_OBJECTS_ACCESSED, OBJECTS_MODIFIED, OBJECT_MODIFIED_BY_DDL, PARENT_QUERY_ID, ROOT_QUERY_ID 
```

Comment:
Space or newline needed here?

Reply:
Apparently not needed. But had include it :slightly_smiling_face: